iota-client/src/api/send.rs

```diff
@@ -330,69 +358,47 @@ impl<'a> SendTransactionBuilder<'a> {
 
         let payload = payload_builder.finish().map_err(|_| Error::TransactionError)?;
 
-        // get tips
-        let tips = self.client.get_tips().await?;
-
         // building message
         let payload = Payload::Transaction(Box::new(payload));
-        let message = MessageBuilder::<ClientMiner>::new()
-            .with_network_id(self.client.get_network_id().await?)
-            .with_parent1(tips.0)
-            .with_parent2(tips.1)
-            .with_payload(payload)
-            .with_nonce_provider(self.client.get_pow_provider(), 4000f64)
-            .finish()
-            .map_err(|_| Error::TransactionError)?;
 
-        self.client.post_message(&message).await
-    }
-}
-
-/// Builder for indexation messages
-pub struct SendIndexationBuilder<'a> {
-    client: &'a Client,
-    index: String,
-    data: Option<Vec<u8>>,
-}
-
-impl<'a> SendIndexationBuilder<'a> {
-    /// Create send builder
-    pub fn new(client: &'a Client, index: String) -> Self {
-        Self {
-            client,
-            index,
-            data: None,
-        }
-    }
-
-    /// Set data to the builder
-    pub fn with_data(mut self, data: Vec<u8>) -> Self {
-        self.data = Some(data);
-        self
+        self.finish_message(Some(payload)).await
     }
 
     /// Consume the builder and get the API result
-    pub async fn finish(self) -> Result<MessageId> {
-        let index = self.index;
-        let data = self.data.unwrap_or_default();
+    pub async fn finish_indexation(self) -> Result<MessageId> {
+        let index = self.index.clone();
+        let data = self.data.clone().unwrap_or_default();
```

Comment:
Why we don't consume them but need to clone them?

Reply:
Because otherwise we can't use self.finish_message() at the end

 ```iota-client\src\api\send.rs:386:9
     |
378  |         let data = self.data.unwrap_or_default();
     |                              ------------------- `self.data` partially moved due to this method call
...
386  |         self.finish_message(Some(payload)).await
     |         ^^^^ value used here after partial move
     |
note: this function consumes the receiver `self` by taking ownership of it, which moves `self.data`
     |
1177 |     pub fn unwrap_or_default(self) -> T {
     |                              ^^^^
     = note: partial move occurs because `self.data` has type `std::option::Option<std::vec::Vec<u8>>`, which does not implement the `Copy` trait```

Reply:
But it seems that we don't need the index and data after `finish_indexation()` Like in the `finish_message()`, we don't need the `index` and `data` anymore. Am I right?

Reply:
Like the following codes, which can build pass. In this way we can avoid the uncessary clone() of the fields.
```rust
pub async fn finish_indexation(self) -> Result<MessageId> {
    let payload: Payload;
    {
        let index = &self.index.as_ref();
        let empty_slice = &vec![];
        let data = &self.data.as_ref().unwrap_or(empty_slice);

        // build indexation
        let index = Indexation::new(index.expect("No indexation tag").to_string(), data)
            .map_err(|e| Error::IndexationError(e.to_string()))?;
        payload = Payload::Indexation(Box::new(index));
    }

    // building message
    self.finish_message(Some(payload)).await
}
```

Reply:
Is this `let empty_slice = &vec![];` required for this to work or works `unwrap_or_default` too?

Reply:
I wanted to use `unwrap_or_default` before. Unfortunately it seems we will need clone() inevitably if we use `unwrap_or_default`

Reply:
The error follows
```
error[E0599]: no method named `unwrap_or_default` found for enum `std::option::Option<&std::vec::Vec<u8>>` in the current scope
   --> iota-client/src/api/send.rs:381:44
    |
381 |             let data = &self.data.as_ref().unwrap_or_default();
    |                                            ^^^^^^^^^^^^^^^^^ method not found in `std::option::Option<&std::vec::Vec<u8>>`
    |
    = note: the method `unwrap_or_default` exists but the following trait bounds were not satisfied:
            `&std::vec::Vec<u8>: std::default::Default`
 ```

Reply:
Ok, fine anyway, will you push it or should I do it?

Reply:
Pushed.